### DISCUSSION TITLE
Custom components changes between spacy 3.1.2 and 3.1.3+

### DISCUSSION THREAD

**Discussion by databill86** [explosion/spaCy#9652] [Help: Coding & Implementations] [answered]
Hello, 

I have a lot of custom components that were perfectly working on spacy==3.1.2, but after upgrading to any 3.1.3+ version, I've got a lot a thrown exceptions: 

> [E893] Could not find function 'my_custom_func' in function registry 'factories'. If you're using a custom function, make sure the code is available. If the function is provided by a third-party package, e.g. spacy-transformers, make sure the package is installed in your environment.
> Available names: 'fr.my_custom_func'...

Each time I try to rerun the build, I get a different 'custom_func' component name. 

Does this have any relation to those two changes made [here](https://github.com/explosion/spaCy/releases/tag/v3.1.3) ?  
> Fix issue #9009: Include component factories in third-party dependencies resolver.
> Fix issue #9012: Correct type of config in create_pipe.

Here is how I registered my components: 

```
""" my_custom_pipelines.py """

@French.factory("my_custom_func", default_config={"language": None,
                                                          ...
                                                          "verbose": False})
def create_custom_func(nlp: Language,
                                  name: str,
                                  language: str,
                                  ....
                                  verbose: bool = False):
    return CustomFunc(nlp, language=language,
                                ....
                                verbose=verbose)
```

Then I save the model to disk:
```
""" my_build_script.py """
import my_custom_pipelines

nlp_extr = spacy.load("fr_core_news_lg") 
nlp_extr.add_pipe("my_custom_func", config={"verbose": False, "language": self.language,
                                                       ....},
                          after="lemmatizer")

nlp_extr.to_disk()
```

And finally I run:

```python -m spacy package models/fr models/mymodel --name=my_model --version=3.9.10 --code=my_custom_pipelines.py```

Am I missing something ? 

Again, there is nothing that I changed between spacy==3.1.2 and 3.1.3+ 

I used _fr_core_news_lg-3.1.0_ for the packaging. 

```python -m spacy download -d fr_core_news_lg-3.1.0```

**Comment by databill86** [comment]
FYI: I restarted everything from a clean environment (spacy 3.1.2), and now, even with spacy==3.1.2, I'm having problems to build the package, I'm getting: 

> atalogue.RegistryError: Cant't find 'spacy.WandbLogger.v1' in registry spacy -> loggers. Available names: spacy-legacy.WandbLogger.v1, spacy.ConsoleLogger.v1, spacy.WandbLogger.v2

Is there a way to disable the WandbLogger ? 

```
[training.logger]
@loggers = "spacy.WandbLogger.v1"
project_name = "spacy-v3.0.0a2"
remove_config_values = []
```

Thanks

**Comment by databill86** [comment, marked as answer]
In the new version 3.2.0, `@Language.factory` works.

**Reply by polm** [reply]
Glad you got it working in 3.2! 

I don't think we were expecting any change in behavior with 3.1.3, but on the other hand since you started having the problem with 3.1.2 too maybe something was weird with your environment. I'll see if I can reproduce the issue.

**Reply by databill86** [reply]
Although, it would be very helpful to see how to make "French.factory" work on s`pacy 3.2.0` and` [fr or es or it]-core-news-lg-3.2.0
`
The other related issue is that on `spacy 3.2.0` and `es-core-news-lg-3.2.0`, I have a custom sentenciser in which I can no longer write to `token.sent_start`. 

`doc[token.i].is_sent_start = False`

The exact same component works perfectly with `spacy 3.2.0` and `es-core-news-lg-3.1.0` (although the latter depends on spacy >=3.1.0,<3.2.0) and does not throw 

> ValueError: [E043] Refusing to write to token.sent_start if its document is parsed, because this may cause inconsistent state.

In both configurations, I have the exact same pipelines and I also excluded `parser` from the model. 

These are the default spacy pipelines: `[.., 'tok2vec', 'morphologizer', 'attribute_ruler', 'lemmatizer',..., 'sentence_segmenter', 'entity_ruler', ...]`

This holds true for french, italian and spanish 'lg' models.

------

Python version: 3.8

**Reply by polm** [reply]
> Although, it would be very helpful to see how to make "French.factory" work on spacy 3.2.0 and[fr or es or it]-core-news-lg-3.2.0

It should just work. I was unable to reproduce your issue with your components not being found. 

It sounds reverting your upgrade didn't make the issue go away, so the problem was something other than the spaCy version. Are you able to recreate the issue in a clean environment?

Regarding the sentence related component, I'm also not sure what's going on there. This code works for me:

```
import spacy
from spacy.language import Language

nlp = spacy.load("es_core_news_lg")

@Language.factory("custom_sent")
def my_senter(nlp, name):
    def default_sents(doc):
        for tok in doc:
            tok.is_sent_start = False
        doc[0].is_sent_start = True
        return doc
    return default_sents

nlp.add_pipe("custom_sent")

doc = nlp("uno dos tres", disable=["parser"])

print(doc)
```

I can only trigger E043 if I remove the `disable` bit.

**Reply by databill86** [reply]
Thank you @polm for your response. 

Actually the original problem was the packaging of model with: `python -m spacy package ...`

I've created a simple project that reproduce the issue: 

https://github.com/databill86/spacy-tests

As you may see in the file src/my_pipelines.py 

```
@French.factory("custom_sent")  # doesn't work for packaging
# @Language.factory("custom_sent")  # work
def my_senter(nlp, name):
    return CustomSenter(nlp)
```

The packaging `python -m spacy package models/my_model models/packaged --name=custom_packaged_spacy --version=0.0.1 --code=src/my_pipelines.py` only work when we use `@Language.factory` instead of `@French.factory`. 

This is the thrown error: 

> catalogue.RegistryError: [E893] Could not find function 'custom_sent' in function registry 'factories'. If you're using a custom function, make sure the code is available. If the function is provided by a third-party package, e.g. spacy-transformers, make sure the package is installed in your environment. 
Available names: attribute_ruler, beam_ner, beam_parser, entity_linker, entity_ruler, fr.custom_sent, fr.lemmatizer, lemmatizer, merge_entities, merge_noun_chunks, merge_subtokens, morphologizer, ner, parser, sentencizer, senter, sp
ancat, tagger, textcat, textcat_multilabel, tok2vec, token_splitter

**Reply by polm** [reply]
Thank you, I didn't realize the issue was `spacy package` specific. Your sample repo illustrates the issue clearly and it looks like a bug. I'll see if I can figure out the details.

**Reply by databill86** [reply]
Thank you!

**Reply by polm** [reply]
Not sure if this is a final fix, but there's a PR in https://github.com/explosion/spaCy/pull/9674 and you can follow development there.

**Reply by databill86** [reply]
This fixes the problem. I hope that's the right fix and that it will be merged soon. 

Thank you again.
